course5/homework02/homework0202.cc

```diff
@@ -0,0 +1,50 @@
+#include <iostream>
+#include <cfloat>
+#include <cmath>
+
+/****
+homework02, part 2
+Write a recursive function that calculates square root using Newton’s method.
+Let the function returns not only the result of the computation, but also some meta data:
+- Depth of the recursions
+- Enough data that allows you to infer the size of a stack frame of the function. There is several ways of doing it; for instance: you can record the address of some local variable in the function in the first and last call of the function, divided it by (depth-1)
+
+Answer the following:
+- which direction does the stack grows?
+- what is the impact of compilation mode (debug/release) on the size of the stack frame
+
+
+Newton's Method, from Wikipedia:
+a method for finding successively better approximations to the roots (or zeroes) of a real-valued function.
+Given a function f and an initial guess x_n at the root, then a better approximation is:
+x_{n+1} = x_{n} - f(x_{n}) / f'(x_{n})
+
+So for square roots:
+f(x) = x^2
+f'(x) = 2x
+x_{n+1} = x_{n} - (x_{n}^2 - x_{n}) / (2 * x_{n})
+***/
+
+float newton_sqrt(float input, float result)
+{
+    if (result * result - input > .00001) {
```

Comment:
I guess this is slightly less arbitrary than just setting the number of iterations beforehand (probably I'm missing something obvious).

But this segfaults if I set the precision too high. If I use `> .000001`, then the segfault occurs trying to calculate the sqrt of 19 (which is 4.3589, according to the output using .00001 as a threshhold). Something to do with floating point precision?